File: dat-go/v4/certificate.go
package dat

import (
	"bytes"
	"strconv"
	"strings"
)

type Certificate struct {
	Cid                     uint64
	cidPreCopy              string
	SignatureKey            *Signature
	CryptoKey               *Crypto
	DatIssuanceStartSeconds uint64
	DatIssuanceEndSeconds   uint64
	DatTtlSeconds           uint64
	expireSeconds           uint64
}

func NewCertificate(cid uint64, datIssuanceStartSeconds, datIssuanceDurationSeconds, datTtlSeconds uint64, signatureKey *Signature, cryptoKey *Crypto) (*Certificate, error) {
	// rust (certificate.rs from) validates the time fields with checked_add only:
	// zero ttl / zero duration are legal, wrapping past u64 is not.
	datIssuanceEndSeconds := datIssuanceStartSeconds + datIssuanceDurationSeconds
	if datIssuanceEndSeconds < datIssuanceStartSeconds {
		return nil, ErrCertMalformed.With("certificate time arithmetic overflowed u64")
	}
	expireSeconds := datIssuanceEndSeconds + datTtlSeconds
	if expireSeconds < datIssuanceEndSeconds {
		return nil, ErrCertMalformed.With("certificate time arithmetic overflowed u64")
	}

	cidPreCopy := "." + ToHexFromU64(cid) + "."

	return &Certificate{
		Cid:                     cid,
		cidPreCopy:              cidPreCopy,
		SignatureKey:            signatureKey,
		CryptoKey:               cryptoKey,
		DatIssuanceStartSeconds: datIssuanceStartSeconds,
		DatIssuanceEndSeconds:   datIssuanceEndSeconds,
		DatTtlSeconds:           datTtlSeconds,
		expireSeconds:           expireSeconds,
	}, nil
}

func GenerateCertificate(cid uint64, datIssuanceStartSeconds, datIssuanceDurationSeconds, datTtlSeconds uint64, signatureAlgorithm SignatureAlgorithm, cryptoAlgorithm CryptoAlgorithm) (*Certificate, error) {
	sk, err := GenerateSignatureKey(signatureAlgorithm)
	if err != nil {
		return nil, err
	}
	ck, err := GenerateCryptoKey(cryptoAlgorithm)
	if err != nil {
		return nil, err
	}
	return NewCertificate(cid, datIssuanceStartSeconds, datIssuanceDurationSeconds, datTtlSeconds, sk, ck)
}

// TryClone deep copies the certificate, so a handed out copy cannot be mutated
// through the manager's live certificate (and vice versa).
func (c *Certificate) TryClone() (*Certificate, error) {
	signatureKey, err := NewSignatureKey(c.SignatureKey.algorithm, bytes.Clone(c.SignatureKey.privateBytes), bytes.Clone(c.SignatureKey.publicBytes))
	if err != nil {
		return nil, err
	}
	cryptoKey, err := NewCryptoKey(c.CryptoKey.algorithm, bytes.Clone(c.CryptoKey.key))
	if err != nil {
		return nil, err
	}
	return &Certificate{
		Cid:                     c.Cid,
		cidPreCopy:              c.cidPreCopy,
		SignatureKey:            signatureKey,
		CryptoKey:               cryptoKey,
		DatIssuanceStartSeconds: c.DatIssuanceStartSeconds,
		DatIssuanceEndSeconds:   c.DatIssuanceEndSeconds,
		DatTtlSeconds:           c.DatTtlSeconds,
		expireSeconds:           c.expireSeconds,
	}, nil
}

func (c *Certificate) Expired() bool {
	return c.expireSeconds < NowUnixTimestamp()
}

func (c *Certificate) Issuable() bool {
	now := NowUnixTimestamp()
	return c.Signable() && now >= c.DatIssuanceStartSeconds && now <= c.DatIssuanceEndSeconds
}

func (c *Certificate) Signable() bool {
	return c.SignatureKey.Signable()
}

func (c *Certificate) SupportVerifyOnly() bool {
	return c.SignatureKey.SupportVerifyOnly()
}

func (c *Certificate) SignatureAlgorithm() SignatureAlgorithm {
	return c.SignatureKey.Algorithm()
}

func (c *Certificate) CryptoAlgorithm() CryptoAlgorithm {
	return c.CryptoKey.Algorithm()
}

func (c *Certificate) Export(verifyOnly bool) (string, error) {
	var sb strings.Builder
	sb.WriteString(ToHexFromU64(c.Cid))
	sb.WriteString(".")
	sb.WriteString(strconv.FormatUint(c.DatIssuanceStartSeconds, 10))
	sb.WriteString(".")
	sb.WriteString(strconv.FormatUint(c.DatIssuanceEndSeconds-c.DatIssuanceStartSeconds, 10))
	sb.WriteString(".")
	sb.WriteString(strconv.FormatUint(c.DatTtlSeconds, 10))
	sb.WriteString(".")
	sb.WriteString(string(c.SignatureKey.Algorithm()))
	sb.WriteString(".")
	sb.WriteString(string(c.CryptoKey.Algorithm()))
	sb.WriteString(".")

	key, err := c.SignatureKey.ExportKeyOption(verifyOnly || !c.Signable())
	if err != nil {
		return "", err
	}

	sb.WriteString(EncodeBase64URL(key))
	sb.WriteString(".")
	sb.WriteString(EncodeBase64URL(c.CryptoKey.ToBytes()))

	return sb.String(), nil
}

func ParseCertificate(format string) (*Certificate, error) {
	parts := strings.Split(format, ".")
	if len(parts) != 8 {
		return nil, ErrCertMalformed.With("expected exactly 8 dot-separated fields")
	}

	cid, err := strconv.ParseUint(parts[0], 16, 64)
	if err != nil {
		return nil, ErrCertMalformed.With("cid field is not a plain hex u64")
	}

	datIssuanceStartSeconds, err := strconv.ParseUint(parts[1], 10, 64)
	if err != nil {
		return nil, ErrCertMalformed.With("certificate field is not a plain decimal u64")
	}

	datIssuanceDurationSeconds, err := strconv.ParseUint(parts[2], 10, 64)
	if err != nil {
		return nil, ErrCertMalformed.With("certificate field is not a plain decimal u64")
	}

	datTtlSeconds, err := strconv.ParseUint(parts[3], 10, 64)
	if err != nil {
		return nil, ErrCertMalformed.With("certificate field is not a plain decimal u64")
	}

	sigAlgo := SignatureAlgorithm(parts[4])
	sigKeyBytes, err := DecodeBase64URL(parts[6])
	if err != nil {
		return nil, err
	}

	signatureKey, err := NewSignatureKey(sigAlgo, sigKeyBytes, nil)
	if err != nil {
		// Try treating as public key if it failed?
		signatureKey, err = NewSignatureKey(sigAlgo, nil, sigKeyBytes)
		if err != nil {
			return nil, err
		}
	}

	cryptoAlgo := CryptoAlgorithm(parts[5])
	cryptoKeyBytes, err := DecodeBase64URL(parts[7])
	if err != nil {
		return nil, err
	}
	cryptoKey, err := NewCryptoKey(cryptoAlgo, cryptoKeyBytes)
	if err != nil {
		return nil, err
	}

	return NewCertificate(cid, datIssuanceStartSeconds, datIssuanceDurationSeconds, datTtlSeconds, signatureKey, cryptoKey)
}

File: dat-rust/src/certificate.rs
use crate::crypto::{DatCrypto, DatCryptoAlgorithm};
use crate::error::DatError;
use crate::signature::{DatSignature, DatSignatureAlgorithm};
use crate::util::{
    decode_base64_url, encode_base64_url_out, now_unix_timestamp, parse_u64_dec, parse_u64_hex,
    to_hex_u64_out,
};
use std::str::FromStr;
use zeroize::{Zeroize, Zeroizing};

pub struct DatCertificate {
    pub cid: u64,
    pub(crate) signature: DatSignature,
    pub(crate) crypto: DatCrypto,
    pub(crate) dat_issuance_start_seconds: u64,
    pub(crate) dat_issuance_end_seconds: u64,
    pub(crate) dat_ttl_seconds: u64,
    pub(crate) expire_seconds: u64,
}

impl DatCertificate {
    pub fn generate(
        cid: u64,
        dat_issuance_start_seconds: u64,
        dat_issuance_duration_seconds: u64,
        dat_ttl_seconds: u64,
        signature_algorithm: DatSignatureAlgorithm,
        crypto_algorithm: DatCryptoAlgorithm,
    ) -> Result<Self, DatError> {
        Self::from(
            cid,
            dat_issuance_start_seconds,
            dat_issuance_duration_seconds,
            dat_ttl_seconds,
            DatSignature::generate(signature_algorithm)?,
            DatCrypto::generate(crypto_algorithm),
        )
    }

    pub fn from(
        cid: u64,
        dat_issuance_start_seconds: u64,
        dat_issuance_duration_seconds: u64,
        dat_ttl_seconds: u64,
        signature_key: DatSignature,
        crypto_key: DatCrypto,
    ) -> Result<Self, DatError> {
        let dat_issuance_end_seconds = dat_issuance_start_seconds
            .checked_add(dat_issuance_duration_seconds)
            .ok_or(DatError::CertMalformed(
                "issuance_start_seconds + issuance_duration_seconds overflowed u64",
            ))?;

        let expire_seconds = dat_issuance_end_seconds.checked_add(dat_ttl_seconds)
            .ok_or(DatError::CertMalformed("issuance_start_seconds + issuance_duration_seconds + dat_ttl_seconds overflowed u64"))?;

        Ok(DatCertificate {
            cid,
            signature: signature_key,
            crypto: crypto_key,
            dat_issuance_start_seconds,
            dat_issuance_end_seconds,
            dat_ttl_seconds,
            expire_seconds,
        })
    }

    #[inline]
    pub fn expired(&self) -> bool {
        self.expire_seconds < now_unix_timestamp()
    }

    #[inline]
    pub fn issuable(&self) -> bool {
        self.signable()
            && (self.dat_issuance_start_seconds..=self.dat_issuance_end_seconds)
                .contains(&now_unix_timestamp())
    }

    #[inline]
    pub fn signable(&self) -> bool {
        self.signature.signable()
    }

    #[inline]
    pub fn support_verify_only(&self) -> bool {
        self.signature.support_verify_only()
    }

    #[inline]
    pub fn signature_algorithm(&self) -> DatSignatureAlgorithm {
        self.signature.algorithm()
    }

    #[inline]
    pub fn crypto_algorithm(&self) -> DatCryptoAlgorithm {
        self.crypto.algorithm()
    }

    pub fn export(&self, verify_only: bool) -> Result<String, DatError> {
        let mut ib = itoa::Buffer::new();
        let mut v: String = String::with_capacity(
            80 + self.signature.key_base64_len() + self.crypto.key_base64_len(),
        );

        to_hex_u64_out(self.cid, &mut v);
        v.push('.');
        v.push_str(ib.format(self.dat_issuance_start_seconds));
        v.push('.');
        v.push_str(ib.format(self.dat_issuance_end_seconds - self.dat_issuance_start_seconds));
        v.push('.');
        v.push_str(ib.format(self.dat_ttl_seconds));
        v.push('.');
        v.push_str(self.signature.algorithm().as_str());
        v.push('.');
        v.push_str(self.crypto.algorithm().as_str());
        v.push('.');
        let mut signature_key = self.signature.export_key_option(verify_only)?;
        encode_base64_url_out(&signature_key, &mut v);
        signature_key.zeroize();

        v.push('.');

        let mut crypto_key = self.crypto.export_key();
        encode_base64_url_out(&crypto_key, &mut v);
        crypto_key.zeroize();

        Ok(v)
    }

    pub fn try_clone(&self) -> Result<Self, DatError> {
        Ok(DatCertificate {
            cid: self.cid,
            signature: self.signature.try_clone()?,
            crypto: self.crypto.clone(),
            dat_issuance_start_seconds: self.dat_issuance_start_seconds,
            dat_issuance_end_seconds: self.dat_issuance_end_seconds,
            dat_ttl_seconds: self.dat_ttl_seconds,
            expire_seconds: self.expire_seconds,
        })
    }
}

impl FromStr for DatCertificate {
    type Err = DatError;
    fn from_str(format: &str) -> Result<Self, Self::Err> {
        let parts = format.split('.').collect::<Vec<&str>>();
        let count = parts.len();
        if count == 8 {
            let cid = parse_u64_hex(parts[0])
                .ok_or(DatError::CertMalformed("cid field is not a plain hex u64"))?;
            let dat_issuance_start_seconds = parse_u64_dec(parts[1]).ok_or(
                DatError::CertMalformed("issuance_start_seconds field is not a plain decimal u64"),
            )?;
            let dat_issuance_duration_seconds =
                parse_u64_dec(parts[2]).ok_or(DatError::CertMalformed(
                    "issuance_duration_seconds field is not a plain decimal u64",
                ))?;
            let dat_ttl_seconds = parse_u64_dec(parts[3]).ok_or(DatError::CertMalformed(
                "dat_ttl_seconds field is not a plain decimal u64",
            ))?;
            let signature_algorithm = DatSignatureAlgorithm::from_str(parts[4])?;
            let crypto_algorithm = DatCryptoAlgorithm::from_str(parts[5])?;
            let signature_key =
                Zeroizing::new(decode_base64_url(parts[6]).map_err(|_| {
                    DatError::CertMalformed("signature_key field is not base64url")
                })?);
            let signature = DatSignature::from_key(signature_algorithm, &signature_key)?;
            let crypto_key = Zeroizing::new(
                decode_base64_url(parts[7])
                    .map_err(|_| DatError::CertMalformed("crypto_key field is not base64url"))?,
            );
            let crypto = DatCrypto::from_key(crypto_algorithm, &crypto_key)?;
            DatCertificate::from(
                cid,
                dat_issuance_start_seconds,
                dat_issuance_duration_seconds,
                dat_ttl_seconds,
                signature,
                crypto,
            )
        } else {
            Err(DatError::CertMalformed(
                "expected exactly 8 dot-separated fields",
            ))
        }
    }
}

impl PartialEq<DatCertificate> for DatCertificate {
    fn eq(&self, other: &DatCertificate) -> bool {
        self.cid.eq(&other.cid)
    }
}

impl PartialEq<u64> for DatCertificate {
    fn eq(&self, other: &u64) -> bool {
        self.cid.eq(other)
    }
}
